File: iosApp/iosApp/p2p/PublicKeyAnnouncementCreator.swift
//
//  PublicKeyAnnouncementCreator.swift
//  iosApp
//
//  Created by jimmyt on 7/29/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import CryptoKit
import Foundation

/**
 Creates a Public Key Announcement for the given public key and offer ID according to the [Commuto Interface Specification](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt).
 
 - Parameters:
    - offerID: The ID of the offer for which this `PublicKey` is being announced.
    - keyPair: The `KeyPair` containing the `PublicKey` to be announced.
 
 - Returns: A JSON `String` that is the Public Key Announcement
 */
func createPublicKeyAnnouncement(offerID: UUID, keyPair: KeyPair) throws -> String {
    // Create Base64-encoded string of public key in PKCS#1 bytes
    let publicKeyString = try keyPair.pubKeyToPkcs1Bytes().base64EncodedString()
    
    // Create a Base64-encoded string of the offer ID
    let offerIDString = offerID.asData().base64EncodedString()
    
    // Create payload Dictionary
    let payload = [
        "pubKey": publicKeyString,
        "offerId": offerIDString
    ]
    
    // Create payload UTF-8 bytes and their Base64-encoded string
    let payloadUTF8Bytes = try JSONSerialization.data(withJSONObject: payload)
    let payloadString = payloadUTF8Bytes.base64EncodedString()
    
    // Create signature of payload hash
    let payloadDataDigest = SHA256.hash(data: payloadUTF8Bytes)
    var payloadDataHashArray = [UInt8]()
    for byte: UInt8 in payloadDataDigest.makeIterator() {
        payloadDataHashArray.append(byte)
    }
    let payloadDataHash = Data(bytes: payloadDataHashArray, count: payloadDataHashArray.count)
    let payloadDataSignature = try keyPair.sign(data: payloadDataHash)
    
    // Create message Dictionary
    let message = [
        "sender": keyPair.interfaceId.base64EncodedString(),
        "msgType": "pka",
        "payload": payloadString,
        "signature": payloadDataSignature.base64EncodedString()
    ]
    
    // Prepare and return message string
    let messageString = String(decoding: try JSONSerialization.data(withJSONObject: message), as: UTF8.self)
    return messageString
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/p2p/create/PublicKeyAnnouncementCreator.kt
package com.commuto.interfacemobile.android.p2p

import com.commuto.interfacemobile.android.key.keys.KeyPair
import com.commuto.interfacemobile.android.key.keys.PublicKey
import com.commuto.interfacemobile.android.p2p.serializable.messages.SerializablePublicKeyAnnouncementMessage
import com.commuto.interfacemobile.android.p2p.serializable.payloads.SerializablePublicKeyAnnouncementPayload
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.security.MessageDigest
import java.util.*

/**
 * Creates a Public Key Announcement for the given public key and offer ID according to the
 * [Commuto Interface Specification](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt).
 *
 * @param offerID The ID of the offer for which the [PublicKey] is being announced.
 * @param keyPair The [KeyPair] containing the [PublicKey] to be announced.
 *
 * @return A JSON [String] that is the Public Key Announcement.
 */
fun createPublicKeyAnnouncement(offerID: UUID, keyPair: KeyPair): String {
    //Setup encoder
    val encoder = Base64.getEncoder()

    // Create Base64-encoded string of public key in PKCS#1 bytes
    val publicKeyString = encoder.encodeToString(keyPair.pubKeyToPkcs1Bytes())

    // Create a Base64-encoded string of the offer ID
    val offerIDByteBuffer = ByteBuffer.wrap(ByteArray(16))
    offerIDByteBuffer.putLong(offerID.mostSignificantBits)
    offerIDByteBuffer.putLong(offerID.leastSignificantBits)
    val offerIDByteArray = offerIDByteBuffer.array()
    val offerIDString = encoder.encodeToString(offerIDByteArray)

    // Create payload object
    val payloadObject = SerializablePublicKeyAnnouncementPayload(
        pubKey = publicKeyString,
        offerId = offerIDString
    )

    // Create payload UTF-8 bytes and their Base64-encoded string
    val payloadString = Json.encodeToString(payloadObject)
    val payloadUTF8Bytes = payloadString.toByteArray(Charset.forName("UTF-8"))

    // Create signature of payload hash
    val payloadDataHash = MessageDigest.getInstance("SHA-256").digest(payloadUTF8Bytes)
    val payloadDataSignature = keyPair.sign(payloadDataHash)

    // Create message object
    val message = SerializablePublicKeyAnnouncementMessage(
        sender = encoder.encodeToString(keyPair.interfaceId),
        msgType = "pka",
        payload = encoder.encodeToString(payloadUTF8Bytes),
        signature = encoder.encodeToString(payloadDataSignature)
    )

    // Prepare and return message string
    return Json.encodeToString(message) // messageString
}